kubernetes/core/src/main/kotlin/com/google/kubernetes/tools/kubectl/KubectlExecutorService.kt

```diff
@@ -0,0 +1,162 @@
+const val KUBECTL_FAIL = "kubectl.fail"
+const val METADATA_ERROR_MESSAGE_KEY = "error.message"
+
+class KubectlExecutorService {
+    companion object {
+        val instance
+            get() = ServiceManager.getService(KubectlExecutorService::class.java)!!
+    }
+
+    var kubectlExecutablePath: String = "kubectl"
+
+    /**
+     * run kubectl version to make sure that kubectl is available, if there's no response after
+     *  a couple of seconds then it is not available.
+     */
+    fun isKubectlAvailable(): Boolean =
+        try {
+            val settings = KubectlExecutorSettings(KubectlExecutorSettings.ExecutionMode.VERSION)
+            val process = executeKubectl(settings).process
+            process.waitFor(1, TimeUnit.SECONDS)
+            process.exitValue() == 0
+        } catch (e: Exception) {
+            false
+        }
+
+    /**
+     * Parse through command line arguments and try to execute the command
+     */
+    fun executeKubectl(settings: KubectlExecutorSettings): KubectlProcess {
+
+        val commandList = mutableListOf<String>()
+
+        with(commandList) {
+            add(kubectlExecutablePath)
+            add(settings.executionMode.modeFlag)
+            settings.executionFlags.forEach {
+                add(it)
+            }
+        }
+
+        try {
+            val processBuilder = ProcessBuilder()
+            val startedProcess = processBuilder.command(commandList).start()
+            val kubectlProcess = KubectlProcess(
+                    startedProcess,
+                    commandList.joinToString(" ")
+            )
+            return kubectlProcess
+        } catch (e: Exception) {
+            val kubectlFailEventName = KUBECTL_FAIL
+
+            UsageTrackerService.getInstance().trackEvent(kubectlFailEventName)
+                    .addMetadata(
+                            METADATA_ERROR_MESSAGE_KEY, e.javaClass.name
+                    ).ping()
+            throw e
+        }
+    }
+
+    /**
+     * Check if kubectl is available, Check if the flags supplied are valid. Create and start the
+     * command line process passed in
+     */
+    fun processOutputToString(executingProcess: Process): String {
+        val reader = BufferedReader(InputStreamReader(executingProcess.inputStream))
+        val builder = StringBuilder()
+        var line = reader.readLine()
+        while (line != null) {
+            builder.append(line)
+            line = reader.readLine()
+            //so that there's no trailing or leading whitespace
+            if (line != null) builder.append(System.getProperty("line.separator"))
+        }
+        return builder.toString()
+    }
+
+    /**
+     * Check if kubectl is available, Check if the flags supplied are valid. Create and start the
+     * command line process passed in. This method waits until the process has been completed and
+     * needs to be run in a separate background thread.
+     */
+    fun startProcess(
```

Comment:
This name sounds a bit generic, can we use something closer to another one in this class, `executeKubectl`? Something along the lines of `waitForKubectlOutput`, or `getKubectlOutputBlocking`?